### PR TITLE
Adding the ability to seed saga data

### PR DESCRIPTION
**Pull request by mat-mcloughlin**
Allows the user to start a testing session providing a seed `IContainsSagaData`. For example:

```
Test.Saga<TestSaga>(new SagaData { Property = "property" })
```

It also includes the ability to "Expect" certain saga data 

```
Test.Saga<TestSaga>(new SagaData  { Property = "property" })
                .ExpectSagaData<SagaData >(data => data.Property == "property")
```

So the user can check their data is in a validate state after calling a handle method.

Connects to #29

### CODE REVIEW

**Comment by mat-mcloughlin** [comment]
@timbussmann @hmemcpy please review

File: src/NServiceBus.Testing.sln.DotSettings [hunk 1]
```diff
@@ -592,6 +592,7 @@ II.2.12 &lt;HandlesEvent /&gt;&#xD;
 	<s:Boolean x:Key="/Default/Environment/SettingsMigration/IsMigratorApplied/=JetBrains_002EReSharper_002EPsi_002ECSharp_002ECodeStyle_002ESettingsUpgrade_002EMigrateBlankLinesAroundFieldToBlankLinesAroundProperty/@EntryIndexedValue">True</s:Boolean>
 	<s:Boolean x:Key="/Default/Environment/SettingsMigration/IsMigratorApplied/=JetBrains_002EReSharper_002EPsi_002ECSharp_002ECodeStyle_002ESettingsUpgrade_002EMigrateThisQualifierSettings/@EntryIndexedValue">True</s:Boolean>
 	<s:Boolean x:Key="/Default/Environment/SettingsMigration/IsMigratorApplied/=JetBrains_002EReSharper_002EPsi_002EJavaScript_002ECodeStyle_002ESettingsUpgrade_002EJsCodeFormatterSettingsUpgrader/@EntryIndexedValue">True</s:Boolean>
+	<s:Boolean x:Key="/Default/Environment/SettingsMigration/IsMigratorApplied/=JetBrains_002EReSharper_002EPsi_002EJavaScript_002ECodeStyle_002ESettingsUpgrade_002EJsWrapperSettingsUpgrader/@EntryIndexedValue">True</s:Boolean>
```
Comment (timbussmann) [inline src/NServiceBus.Testing.sln.DotSettings]: what's this?

File: src/NServiceBus.Testing/Saga.cs [hunk 2]
```diff
@@ -468,5 +477,6 @@ void HandleTimeout<TMessage>(TMessage message)
         TestableMessageHandlerContext testableMessageHandlerContext;
 
         Lazy<MethodInfo> timeoutMethodInfo = new Lazy<MethodInfo>(() => typeof(Saga<T>).GetMethod("HandleTimeout", BindingFlags.Instance | BindingFlags.NonPublic));
+
```
Comment (timbussmann) [inline src/NServiceBus.Testing/Saga.cs]: :fire:

File: src/NServiceBus.Testing/Test.cs [hunk 3]
```diff
@@ -25,25 +25,33 @@ public static Saga<TSaga> Saga<TSaga>() where TSaga : Saga, new()
             return new Saga<TSaga>(saga);
         }
 
-        /// <summary>
+        ///<summary>
         /// Begin the test script for a saga of type T while specifying the saga id.
-        /// </summary>
-        public static Saga<T> Saga<T>(Guid sagaId) where T : Saga, new()
+        ///</summary>
+        public static Saga<TSaga> Saga<TSaga>(Guid sagaId) where TSaga : Saga
```
Comment (timbussmann) [inline src/NServiceBus.Testing/Test.cs]: why did you remove the `new()` constraint?

File: src/NServiceBus.Testing/Test.cs [hunk 4]
```diff
@@ -20,30 +20,38 @@ public static Saga<TSaga> Saga<TSaga>() where TSaga : Saga, new()
         /// Begin the test script for the passed in saga instance.
         /// Callers need to instantiate the saga's data class as well as give it an ID.
         /// </summary>
-        public static Saga<TSaga> Saga<TSaga>(TSaga saga) where TSaga : Saga
+        public static Saga<TSaga> Saga<TSaga>(TSaga saga) where TSaga : Saga, new()
```
Comment (timbussmann) [inline src/NServiceBus.Testing/Test.cs]: we removed the new constrain here since the user already provides the saga instance. So I'm :-1: for adding it again.

File: src/NServiceBus.Testing/Test.cs [hunk 5]
```diff
@@ -20,30 +20,38 @@ public static Saga<TSaga> Saga<TSaga>() where TSaga : Saga, new()
...
-                if (saga.Entity != null)
-                {
-                    saga.Entity.Id = sagaId;
-                }
+        /// <summary>
+        /// Begin the test script for a saga of type T with the passed in in <see cref="IContainSagaData"/>.
+        /// </summary>
+        public static Saga<TSaga> Saga<TSaga>(IContainSagaData sagaData) where TSaga : Saga, new()
+        {
+            var saga = (TSaga)Activator.CreateInstance(typeof(TSaga));
```
Comment (timbussmann) [inline src/NServiceBus.Testing/Test.cs]: can't we just call `new TSaga()`?

File: src/NServiceBus.Testing/Test.cs [hunk 6]
```diff
@@ -20,30 +20,38 @@ public static Saga<TSaga> Saga<TSaga>() where TSaga : Saga, new()
...
+
+            if (sagaData != null)
```
Comment (timbussmann) [inline src/NServiceBus.Testing/Test.cs]: can that ever be null?

File: src/NServiceBus.Testing/Saga.cs [hunk 7]
```diff
@@ -236,7 +236,7 @@ public Saga<T> WhenHandling<TMessage>(Action<TMessage> initializeMessage = null)
         {
             var message = messageCreator.CreateInstance(initializeMessage);
 
-            return When((s, c) => ((dynamic) s).Handle(message, c));
+            return When((s, c) => ((dynamic)s).Handle(message, c));
```
Comment (timbussmann) [inline src/NServiceBus.Testing/Saga.cs]: I think you need to run a code reformat on that file again

File: src/NServiceBus.Testing.Tests/Saga/SagaDataTests.cs [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+        }
+
+        [Test]
+        public void ShouldUpdateSagaDataWhenChangeMessageIsCalled()
+        {
+            Test.Saga<SagaCreation>(new SagaCreation.CreationData { Property = "property" })
+                .ExpectSagaData<SagaCreation.CreationData>(data => data.Property == "newData")
+                .WhenHandling<SagaCreation.ChangeMessage>();
+        }
+    }
```
Comment (timbussmann) [inline src/NServiceBus.Testing.Tests/Saga/SagaDataTests.cs]: It may make sense to have a test to verify an exception is thrown when the expectation isn't met?
